python/demo/demo_scipy-solvers.py

```diff
@@ -0,0 +1,156 @@
+# project the function `g` into.
+
+Nx = 15
+Ny = 20
+domain = dolfinx.mesh.create_unit_square(MPI.COMM_WORLD, Nx, Ny)
+V = dolfinx.fem.functionspace(domain, ("Lagrange", 1))
+x, y = ufl.SpatialCoordinate(domain)
+g = ufl.sin(ufl.pi * x) * ufl.cos(ufl.pi * y)
+
+# The functional $G$ can be defined as follows
+
+uh = dolfinx.fem.Function(V)
+G = ufl.dot(uh - g, uh - g) * ufl.dx
+
+G_compiled = dolfinx.fem.form(G)
+
+# We can find the minimum of this function by differentiating
+# the function with respect to `u`:
+#
+# $$ F(u, \delta u) = \frac{\mathrm{d}G}{\mathrm{d}u}[\delta u] = 0, \quad \forall \delta u \in V.$$
+
+du = ufl.conj(ufl.TestFunction(V))
+residual = ufl.derivative(G, uh, du)
+
+# We generate the integration kernels for the residual, and assemble an initial
+# residual vector `b`.
+
+F = dolfinx.fem.form(residual)
+b = dolfinx.fem.assemble_vector(F)
+
+# A convenience function for computing the value of the functional
+
+
+def compute_functional_value() -> dolfinx.default_scalar_type:
+    local_contribution = dolfinx.fem.assemble_scalar(G_compiled)
+    return domain.comm.allreduce(local_contribution, op=MPI.SUM)
+
+
+# We can use the scipy Newton-Krylov solver to solve this problem.
+# This solver takes in a function that computes the residual of the functional
+# We re-use our structures for `uh` and `b` to avoid unnecessary allocations.
+
+
+def compute_residual(x) -> npt.NDArray[dolfinx.default_scalar_type]:
+    """
+    Evaluate the residual F(x) = 0
+
+    Args:
+        x: Input vector with current solution
+    Returns:
+        Residual array
+    """
+    uh.x.array[:] = x
+    b.array[:] = 0
+    dolfinx.fem.assemble_vector(b.array, F)
+    b.scatter_reverse(dolfinx.la.InsertMode.add)
+    return b.array
+
+
+uh.x.array[:] = 0
+print("Jacobian free Newton Krylov")
+print(f"Initial functional {compute_functional_value():.2e}")
+solution = scipy.optimize.newton_krylov(compute_residual, uh.x.array, method="lgmres", verbose=True)
+uh.x.array[:] = solution
+print(f"Minimal functional {compute_functional_value():.2e}")
+
+# We could also exploit the Jacobian information and use a Newton method with an exact gradient
+#
+#  $$ J(u, \delta u, \delta v) = \frac{\mathrm{d} F}{\mathrm{d}u}[\delta u, \delta v] =
+# \frac{\mathrm{d}^2G}{\mathrm{d}u^2}[\delta u, \delta v].$$
+
+dv = ufl.TrialFunction(V)
+jacobian = ufl.derivative(residual, uh, dv)
+
+# We assemble the Jacobian matrix and its inverse.
+# Note that we assemble the Jacobian into a :class:`dolfinx.la.MatrixCSR`.
+# Then, we get a view of the data in this matrix by using
+# :func:`dolfinx.la.MatrixCSR.to_scipy`.
+
+J_compiled = dolfinx.fem.form(jacobian)
+A = dolfinx.fem.assemble_matrix(J_compiled)
+A_scipy = A.to_scipy()
+Ainv = scipy.sparse.linalg.splu(A_scipy)
+
+# If the Jacobian is independent of uh, we can avoid re-assembling the matrix
+
+is_nonlinear = uh in ufl.algorithms.expand_derivatives(jacobian).coefficients()
```

Comment:
@jhale I've here used `ufl to figure it out.
I think it is a quite nice point to make, as some people prefer always using a non-linear solver, but could benefit alot from not having to reassemble J at every step of the Newton solver (even if it is going to be just 1).